govcd/vdc.go

```diff
@@ -901,6 +901,169 @@ func (vdc *Vdc) GetVappList() []*types.ResourceReference {
+// It retrieves the VM owner (the hidden vApp), and from that one finds the new VM
+func (vdc *Vdc) getVmFromTask(task Task, name string) (*VM, error) {
+	owner := task.Task.Owner.HREF
+	if owner == "" {
+		return nil, fmt.Errorf("task owner is null for VM %s", name)
+	}
+	vapp, err := vdc.GetVAppByHref(owner)
+	if err != nil {
+		return nil, err
+	}
+	if vapp.VApp.Children == nil {
+		return nil, ErrorEntityNotFound
+	}
+	if len(vapp.VApp.Children.VM) == 0 {
+		return nil, fmt.Errorf("vApp %s contains no VMs", vapp.VApp.Name)
+	}
+	if len(vapp.VApp.Children.VM) > 1 {
+		return nil, fmt.Errorf("vApp %s contains more than one VM", vapp.VApp.Name)
+	}
+	for _, child := range vapp.VApp.Children.VM {
+		util.Logger.Printf("[TRACE] Looking at: %s", child.Name)
+		return vapp.client.GetVMByHref(child.HREF)
+	}
+	return nil, ErrorEntityNotFound
+}
+
+// CreateStandaloneVm creates a standalone VM without a template
+func (vdc *Vdc) CreateStandaloneVm(params *types.CreateVmParams) (*VM, error) {
+
+	task, err := vdc.CreateStandaloneVmAsync(params)
+	if err != nil {
+		return nil, err
+	}
+	err = task.WaitTaskCompletion()
+	if err != nil {
+		return nil, err
+	}
+	return vdc.getVmFromTask(task, params.Name)
+}
+
+// QueryVmByName finds a standalone VM by name
+// The search fails either if there are more VMs with the wanted name, or if there are none
+// It can also retrieve a standard VM (created from vApp)
+func (vdc *Vdc) QueryVmByName(name string) (*VM, error) {
+	vmList, err := vdc.QueryVmList(types.VmQueryFilterOnlyDeployed)
+	if err != nil {
+		return nil, err
+	}
+	var foundVM []*types.QueryResultVMRecordType
+	for _, vm := range vmList {
+		if vm.Name == name {
+			foundVM = append(foundVM, vm)
+		}
+	}
+	if len(foundVM) == 0 {
+		return nil, ErrorEntityNotFound
+	}
+	if len(foundVM) > 1 {
+		return nil, fmt.Errorf("more than one VM found with name %s", name)
+	}
+	return vdc.client.GetVMByHref(foundVM[0].HREF)
+}
+
+// QueryVmById retrieves a standalone VM by ID
+// It can also retrieve a standard VM (created from vApp)
+func (vdc *Vdc) QueryVmById(id string) (*VM, error) {
+	vmList, err := vdc.QueryVmList(types.VmQueryFilterOnlyDeployed)
+	if err != nil {
+		return nil, err
+	}
+	var foundVM []*types.QueryResultVMRecordType
+	for _, vm := range vmList {
+		if equalIds(id, vm.ID, vm.HREF) {
+			foundVM = append(foundVM, vm)
+		}
+	}
+	if len(foundVM) == 0 {
+		return nil, ErrorEntityNotFound
+	}
+	if len(foundVM) > 1 {
+		return nil, fmt.Errorf("more than one VM found with ID %s", id)
+	}
+	return vdc.client.GetVMByHref(foundVM[0].HREF)
+}
+
+// CreateStandaloneVMFromTemplateAsync starts a standalone VM creation using a template
+func (vdc *Vdc) CreateStandaloneVMFromTemplateAsync(params *types.InstantiateVmTemplateParams) (Task, error) {
+
+	util.Logger.Printf("[TRACE] Vdc.CreateStandaloneVMFromTemplateAsync - Creating VM ")
+
+	if vdc.Vdc.HREF == "" {
+		return Task{}, fmt.Errorf("cannot create VM, Object VDC is empty")
```

Comment:
Do we need "Object"? Maybe:

```suggestion
		return Task{}, fmt.Errorf("cannot create VM, provided VDC is empty")
```

Reply:
changed